pkg/volume/emptydir/empty_dir.go

```diff
@@ -331,10 +331,10 @@ func (ed *emptyDir) setupTmpfs(dir string) error {
 		return nil
 	}
 
-	var options []string
+	options := []string{"noswap"}
```

Comment:
Looks like 6.4 onwards only!
- https://github.com/torvalds/linux/commit/2c6efe9cf2d7841b75fe38ed1adbd41a90f51ba0
- https://lwn.net/ml/linux-kernel/20230309230545.2930737-7-mcgrof@kernel.org/

Reply:
Questions for node approvers:
1. What other things do we have in the kubelet that assume / require minimum kernel versions?
2. How do we detect and document those minimum requirements?
3. What do we do when the kernel doesn't meet those requirements? (error early on startup, error late at runtime, non-error fail open at runtime, something else?)

Reply:
Thanks a lot @liggitt and @dims!

That's very helpful. I'll definitely look into it.

Reply:
>What other things do we have in the kubelet that assume / require minimum kernel versions?

When using a metric `container_memory_max_usage_bytes` in cgroup v2, the minimum kernel required kernel is 5.19, otherwise that metric will return 0.

xref: https://github.com/kubernetes/kubernetes/pull/122664#issuecomment-1899581854

Reply:
A minimum kernel version to have specific metrics available is a good data point. Do we have examples where a minimum kernel version is required for a container or volume to operate as expected?

Reply:
minimum kernel version is tricky though because downstreams backport patches to older kernel versions. having a way to probe the kernel for a specific feature is more reliable (like doing a dummy mount at kubelet startup to see if tmpfs supports noswap)

Reply:
Probing at start sounds great if we can make it reliable. I guess we would only need to do this if swap was enabled.

The next question is what the kubelet should do if swap is enabled, and the kubelet is configured to run with swap on, and the kernel version doesn't allow `noswap` for tmpfs. I *think* the kubelet should fail safe and refuse to run.

Reply:
> The next question is what the kubelet should do if swap is enabled, and the kubelet is configured to run with swap on, and the kernel version doesn't allow noswap for tmpfs. I think the kubelet should fail safe and refuse to run.

Yes, that sounds good.

Reply:
It may be difficult to enforce that last point as it seems that developer environments (hack/local-up-cluster enables swap by default) may use swap. Since tmpfs swap is the default option, I think that limits our options.

I know that when I tried to fail swap on cgroup v1 I ran into quite a few tests/ci environments that failed on that which was surprising to me.

I think "swap is enabled" will have to mean `--fail-swap-on=false` in this case as this is not really limited to pods swapping but whether or not kubelet won't fail on swap.

At a quick search, I did see some k8s components that set to this to false. https://cs.k8s.io/?q=failSwapOn%3A%20false&i=nope&files=&excludeFiles=&repos=

Reply:
Looking at these test logs..

```
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074116     745 mount_linux.go:224] Mounting cmd (mount) with arguments (-t tmpfs -o noswap,size=54736998400 tmpfs /var/lib/kubelet/pods/75271a50-bac8-4506-8961-cf8331019a18/volumes/kubernetes.io~projected/kube-api-access-tvkpq)
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074051     745 empty_dir_linux.go:99] Statfs_t of /var/lib/kubelet/pods/45452852-ccfd-4739-8693-6e216c5f31cb/volumes/kubernetes.io~projected/kube-api-access-pzjxq: {Type:61267 Bsize:4096 Blocks:193176694 Bfree:185170145 Bavail:175338952 Files:49135616 Ffree:48642004 Fsid:{Val:[1454556786 880505981]} Namelen:255 Frsize:4096 Flags:4128 Spare:[0 0 0 0]}
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074213     745 empty_dir.go:340] pod 45452852-ccfd-4739-8693-6e216c5f31cb: mounting tmpfs for volume wrapped_kube-api-access-pzjxq
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074234     745 mount_linux.go:224] Mounting cmd (mount) with arguments (-t tmpfs -o noswap,size=52428800 tmpfs /var/lib/kubelet/pods/45452852-ccfd-4739-8693-6e216c5f31cb/volumes/kubernetes.io~projected/kube-api-access-pzjxq)
Mar 26 15:10:10 kind-control-plane kubelet[745]: E0326 15:10:10.081410     745 mount_linux.go:236] Mount failed: exit status 32
```

Looks like we got the behavior we want if the feature is not set. We get an exit code of 32 when mounting on startup. Kind uses fail-swap-on=false by default I guess.

Reply:
@kannon92 

> Looks like we got the behavior we want if the feature is not set.

Yes, that's true, in the current implementation the `noswap` option is always being added, not only when using `--fail-swap-on=false`. However I agree with @liggitt that it sounds like the right path forward. I'll work on that.

> We get an exit code of 32 when mounting on startup. 

Do we know which kernel version is used in k8s CI? Is it >=6.4?

> Probing at start sounds great if we can make it reliable.

I fully agree. I'll try to understand how / if it's impossible to do so.

Reply:
> Do we know which kernel version is used in k8s CI? Is it >=6.4?

Its difficult to answer that question. 

On the pull-e2e-kind tests, I see `Mar 26 15:09:49 kind-control-plane kubelet[250]: I0326 15:09:49.330463     250 manager.go:233] Version: {KernelVersion:5.15.0-1048-gke ContainerOsVersion:Debian GNU/Linux 12 (bookworm) DockerVersion: DockerAPIVersion: CadvisorVersion: CadvisorRevision:}
`

So I can see that we using debian 12 and supposely the kernel is 5.15? But I think there are backports. @haircommander mentioned to me that using kernel version is pretty much impossible with all these back ports.

Reply:
@pacoxu has this issue https://github.com/kubernetes/kubernetes/issues/116799 where they mention features and kernel support.

Reply:
Is there any chance that we make this as a mount opts? I know this is KEP focusing on swap thing, but we might hope to custumize more opts like huge=advise for THP, or readOnly kind of opts

Reply:
I'm not sure if you're proposing exposing mount options via the API, but those are not generally safe to expose to end users. The intent of this is not to expose more levers and knobs, but to make the existing API surface continue to operate safely even when swap is enabled.

Reply:
yes, I'm proposing exposing mount Options. But also agree with your concern here, like noexec might bring unexpected security issue

Reply:
but how about making a subset as mount Options?